### ISSUE TITLE
Tab for spaces proj and global setting doesn't seem to followed in Rmd vignette files

### ISSUE THREAD

**Issue by lcolladotor** [rstudio/rstudio#5454]
Hi,

I've encountered an "insert spaces for tab" issue that might be similar to https://github.com/rstudio/rstudio/issues/4448 and https://github.com/rstudio/rstudio/issues/4454.

## Steps that work well

* Use `RStudio`'s menus to create a new empty project called `testproj`.

<img width="531" alt="Screen Shot 2019-09-24 at 11 52 33 AM" src="https://user-images.githubusercontent.com/2288213/65528779-02e40e00-dec3-11e9-9944-c12042a98814.png">

* Use the `RStudio`'s menus to create a new R Markdown file.

<img width="525" alt="Screen Shot 2019-09-24 at 11 53 00 AM" src="https://user-images.githubusercontent.com/2288213/65528817-13948400-dec3-11e9-9381-16702b161f08.png">

* Set/check that "insert spaces for tab" is set to 4 in the project options.

<img width="542" alt="Screen Shot 2019-09-24 at 11 53 16 AM" src="https://user-images.githubusercontent.com/2288213/65528876-28711780-dec3-11e9-9478-50ffcd4448b4.png">

* and the global options.

<img width="591" alt="Screen Shot 2019-09-24 at 11 53 38 AM" src="https://user-images.githubusercontent.com/2288213/65528897-31fa7f80-dec3-11e9-9d1e-0f7ef350917a.png">

* You can start typing in a code chunk and the 4 spaces will be respected as show below

<img width="213" alt="Screen Shot 2019-09-24 at 11 55 06 AM" src="https://user-images.githubusercontent.com/2288213/65528917-4179c880-dec3-11e9-9c45-fa58dcddd9f5.png">

* Or you can type and delete/change the indentation manually as below

<img width="192" alt="Screen Shot 2019-09-24 at 11 56 05 AM" src="https://user-images.githubusercontent.com/2288213/65528966-522a3e80-dec3-11e9-9ddc-b926ac32e2f8.png">

and then use `cmd + i` (MacOS) to auto-indent the code and return to the correct look with 4 spaces.

<img width="202" alt="Screen Shot 2019-09-24 at 11 56 13 AM" src="https://user-images.githubusercontent.com/2288213/65529021-65d5a500-dec3-11e9-99b3-722be46e12c4.png">

## Steps that don't work

* Create a new package using `usethis::create_package()`.

```R
> usethis::create_package('~/Dropbox/Code/testpkg')
✔ Creating '/Users/lcollado/Dropbox/Code/testpkg/'
✔ Setting active project to '/Users/lcollado/Dropbox/Code/testpkg'
✔ Creating 'R/'
✔ Writing 'DESCRIPTION'
Package: testpkg
Title: What the Package Does (One Line, Title Case)
Version: 0.0.0.9000
Authors@R (parsed):
    * First Last <first.last@example.com> [aut, cre] (<https://orcid.org/YOUR-ORCID-ID>)
Description: What the package does (one paragraph).
License: What license it uses
Encoding: UTF-8
LazyData: true
✔ Writing 'NAMESPACE'
✔ Writing 'testpkg.Rproj'
✔ Adding '.Rproj.user' to '.gitignore'
✔ Adding '^testpkg\\.Rproj$', '^\\.Rproj\\.user$' to '.Rbuildignore'
✔ Opening '/Users/lcollado/Dropbox/Code/testpkg/' in new RStudio session
✔ Setting active project to '<no active project>'
```

* Check that the project settings has 4 spaces per tab selected.

<img width="546" alt="Screen Shot 2019-09-24 at 11 57 42 AM" src="https://user-images.githubusercontent.com/2288213/65529119-8998eb00-dec3-11e9-95d1-9321c3404983.png">

* Create a new R Markdown vignette file.

```R
> usethis::use_vignette('test-vignette')
✔ Setting active project to '/Users/lcollado/Dropbox/Code/testpkg'
✔ Adding 'knitr' to Suggests field in DESCRIPTION
✔ Setting VignetteBuilder field in DESCRIPTION to 'knitr'
✔ Adding 'inst/doc' to '.gitignore'
✔ Creating 'vignettes/'
✔ Adding '*.html', '*.R' to 'vignettes/.gitignore'
✔ Adding 'rmarkdown' to Suggests field in DESCRIPTION
✔ Writing 'vignettes/test-vignette.Rmd'
● Modify 'vignettes/test-vignette.Rmd'
```

* Try to use `cmd + i` to auto-indent the start of the vignette that includes a code chunk with 2 spaces instead of 4. It doesn't change.

<img width="210" alt="Screen Shot 2019-09-24 at 11 57 56 AM" src="https://user-images.githubusercontent.com/2288213/65529190-ae8d5e00-dec3-11e9-9b56-2208d83388be.png">

* Try to type in a new code chunk. It still uses 2 spaces instead of 4.

<img width="169" alt="Screen Shot 2019-09-24 at 11 58 35 AM" src="https://user-images.githubusercontent.com/2288213/65529233-c1a02e00-dec3-11e9-9239-7dd4d2d60964.png">

I've encountered this issue in several of my packages (`derfinder`, `recount` from Bioc, `LieberInstitute/sgejobs`, `LieberInstitute/libdRSE` from GitHub to name a few), regardless of whether they were created with `usethis::create_package()` and regardless of how the R Markdown vignette was made (`usethis::use_vignette()` or other ways). I've also seen this with my Windows laptop (I can add the info for that setup later if needed, I don't have it here with me right now).

Of note, I just tried creating a new R Markdown file at `testpkg` using the RStudio menus and that file (not saved inside `vignettes') does use 4 spaces per tab. So maybe it's something specific to vignette files.

Changing this portion of the YAML front matter in the vignette

```
vignette: >
  %\VignetteIndexEntry{test-vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
```

to

```
vignette: >
    %\VignetteIndexEntry{test-vignette}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
```

had no effect (2 to 4 spaces) in the indentation of the R code further below. 

So I don't have any more ideas on what to try to further trace this issue. If you need any more information, please let me know.

Best,
Leo

### `testpkg/vignettes/test-vignette.Rmd` contents

````
---
title: "test-vignette"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{test-vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(testpkg)
```

```{r}
df <- data.frame(
  a = 'hola'
)
```
````

### `testpkg/Untitled.Rmd` contents

````
---
title: "Test Rmd 2"
author: "L. Collado-Torres"
date: "9/24/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
summary(cars)
```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.

```{r}
data.frame(
    'a' = hola
)
```
````

Note the two spaces before `'a' = hola` instead of four.

### RStudio (macOS laptop)

```
RStudio
Version 1.2.1522
© 2009-2019 RStudio, Inc.
Build unknown (170d44fa, 2019-05-17)
Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) QtWebEngine/5.12.1 Chrome/69.0.3497.128 Safari/537.36
```

### R session

```R
> library(sessioninfo)
> options(width = 120)
> session_info()
─ Session info ───────────────────────────────────────────────────────────────────────────────────────────────────────
 setting  value                       
 version  R version 3.6.1 (2019-07-05)
 os       macOS Mojave 10.14.6        
 system   x86_64, darwin15.6.0        
 ui       RStudio                     
 language (EN)                        
 collate  en_US.UTF-8                 
 ctype    en_US.UTF-8                 
 tz       America/New_York            
 date     2019-09-24                  

─ Packages ───────────────────────────────────────────────────────────────────────────────────────────────────────────
 package     * version date       lib source                            
 assertthat    0.2.1   2019-03-21 [1] CRAN (R 3.6.0)                    
 backports     1.1.4   2019-04-10 [1] CRAN (R 3.6.0)                    
 callr         3.3.2   2019-09-22 [1] CRAN (R 3.6.1)                    
 cli           1.1.0   2019-03-19 [1] CRAN (R 3.6.0)                    
 clisymbols    1.2.0   2017-05-21 [1] CRAN (R 3.6.0)                    
 colorout    * 1.2-1   2019-05-07 [1] Github (jalvesaq/colorout@7ea9440)
 crayon        1.3.4   2017-09-16 [1] CRAN (R 3.6.0)                    
 desc          1.2.0   2018-05-01 [1] CRAN (R 3.6.0)                    
 devtools    * 2.2.0   2019-09-07 [1] CRAN (R 3.6.0)                    
 digest        0.6.21  2019-09-20 [1] CRAN (R 3.6.0)                    
 DT            0.9     2019-09-17 [1] CRAN (R 3.6.0)                    
 ellipsis      0.3.0   2019-09-20 [1] CRAN (R 3.6.0)                    
 evaluate      0.14    2019-05-28 [1] CRAN (R 3.6.0)                    
 fs            1.3.1   2019-05-06 [1] CRAN (R 3.6.0)                    
 git2r         0.26.1  2019-06-29 [1] CRAN (R 3.6.0)                    
 glue          1.3.1   2019-03-12 [1] CRAN (R 3.6.0)                    
 htmltools     0.3.6   2017-04-28 [1] CRAN (R 3.6.0)                    
 htmlwidgets   1.3     2018-09-30 [1] CRAN (R 3.6.0)                    
 knitr         1.25    2019-09-18 [1] CRAN (R 3.6.0)                    
 magrittr      1.5     2014-11-22 [1] CRAN (R 3.6.0)                    
 memoise       1.1.0   2017-04-21 [1] CRAN (R 3.6.0)                    
 packrat       0.5.0   2018-11-14 [1] CRAN (R 3.6.0)                    
 pkgbuild      1.0.5   2019-08-26 [1] CRAN (R 3.6.1)                    
 pkgload       1.0.2   2018-10-29 [1] CRAN (R 3.6.0)                    
 prettyunits   1.0.2   2015-07-13 [1] CRAN (R 3.6.0)                    
 processx      3.4.1   2019-07-18 [1] CRAN (R 3.6.0)                    
 ps            1.3.0   2018-12-21 [1] CRAN (R 3.6.0)                    
 R6            2.4.0   2019-02-14 [1] CRAN (R 3.6.0)                    
 Rcpp          1.0.2   2019-07-25 [1] CRAN (R 3.6.0)                    
 remotes       2.1.0   2019-06-24 [1] CRAN (R 3.6.0)                    
 rlang         0.4.0   2019-06-25 [1] CRAN (R 3.6.0)                    
 rmarkdown     1.15    2019-08-21 [1] CRAN (R 3.6.1)                    
 rprojroot     1.3-2   2018-01-03 [1] CRAN (R 3.6.0)                    
 rsconnect     0.8.15  2019-07-22 [1] CRAN (R 3.6.0)                    
 rstudioapi    0.10    2019-03-19 [1] CRAN (R 3.6.0)                    
 sessioninfo * 1.1.1   2018-11-05 [1] CRAN (R 3.6.0)                    
 testthat    * 2.2.1   2019-07-25 [1] CRAN (R 3.6.0)                    
 usethis     * 1.5.1   2019-07-04 [1] CRAN (R 3.6.0)                    
 whisker       0.4     2019-08-28 [1] CRAN (R 3.6.0)                    
 withr         2.1.2   2018-03-15 [1] CRAN (R 3.6.0)                    
 xfun          0.9     2019-08-21 [1] CRAN (R 3.6.0)                    

[1] /Library/Frameworks/R.framework/Versions/3.6/Resources/library
```

**Comment by kevinushey** [comment]
Do you have "[ ] Auto-detect code indentation" enabled?

![Screen Shot 2019-09-24 at 10 19 56 AM](https://user-images.githubusercontent.com/1976582/65534770-dfb26200-deb4-11e9-8ceb-99eb62ff8e28.png)

If so, does disabling that and trying once more help?

I suspect the issue here could be that our template file comes with code indented with two spaces, and our auto-detect engine sees that and overrides the default project settings.

I think we need to transform the indentation used in our template files according to the indentation used in the project.

tl;dr: It looks like this is a bug in RStudio. Thanks for reporting!

**Comment by lcolladotor** [comment]
Hi,

I did have `auto-detect code indentation` enabled (see global settings screenshot above). After disabling it, the auto-indentation didn't work with `testpkg/vignettes/test-vignette.Rmd`. However, after closing `testpkg/vignettes/test-vignette.Rmd` and re-opening it (without closing RStudio), the indentation (4 spaces per tab) works correctly 🎉

Though I guess that based on the mouse over tooltip for `auto-detect code indentation`, this is still a bug (since I'm inside an RStudio project) 🤔.

<img width="584" alt="Screen Shot 2019-09-24 at 1 38 04 PM" src="https://user-images.githubusercontent.com/2288213/65536046-9c65ec80-ded0-11e9-9706-3312fade0550.png">

Thank you for your help! The source of the issue that you described makes lots of sense to me. Maybe if you could read the setting then run something like `formatR` or whatever is run by `cmd + i` on the templates, then it'll work, no? From https://github.com/rstudio/rstudio/issues/1607 and https://github.com/rstudio/rstudio/pull/5069 I get the impression that reading that particular global setting will be possible soon.

Best,
Leo

**Comment by vadimus202** [comment]
Thanks everyone for detailed background info on this. I have existing Rmd files with YAML headers indented with 2 spaces, and it's been driving me crazy. Re-indenting YAML also solves the issue.

**Comment by lc822** [comment]
Thanks everyone - I have the same issue.  So unticking `Auto-detect code indentation` should do the trick?  Or has this bug been fixed?

**Comment by mjsteinbaugh** [comment]
This bug still persists in RStudio 1.4.1103.

**Comment by mpettis** [comment]
Same.  It seems to switch randomly between 2- and 4- space indents, without me knowingly changing any global or project options.  So, same problem here.

**Comment by kevinushey** [comment]
Have you confirmed whether auto-detect code indentation is enabled? Does disabling it make a difference?

A reproducible example would be helpful.

**Comment by mpettis** [comment]
Thanks @kevinushey .  I'm not sure how to do reproducible examples for configuration stuff (I know how to do this for code with reprexes).  But I'll say that I followed all of the steps of the original post, and did restart the session between.  I do know I have an older version of RStudio (1.3.1093), so the recommendation might be to update that.  Beyond that, I'll take guidance.  My screenshots look identical to the ones of the OP.
